massa-module-cache/src/controller.rs

```diff
@@ -0,0 +1,159 @@
+// IMPORTANT TODO: change bc update ABI costs in abi_gas_costs.json
+// IMPORTANT TODO: move non-error warn logs to debug
+
+use massa_hash::Hash;
+use massa_models::prehash::BuildHashMapper;
+use massa_sc_runtime::{Compiler, RuntimeModule};
+use schnellru::{ByLength, LruMap};
+use tracing::warn;
+
+use crate::{
+    config::ModuleCacheConfig, error::CacheError, hd_cache::HDCache, lru_cache::LRUCache,
+    types::ModuleInfo,
+};
+
+/// `LruMap` specialization for `PreHashed` keys
+pub type PreHashLruMap<K, V> = LruMap<K, V, ByLength, BuildHashMapper<K>>;
+
+/// Cache controller of compiled runtime modules
+pub struct ModuleCache {
+    /// Cache config.
+    /// See `CacheConfig` documentation for more information.
+    cfg: ModuleCacheConfig,
+    /// RAM stored LRU cache.
+    /// See `LRUCache` documentation for more information.
+    lru_cache: LRUCache,
+    /// Disk stored cache.
+    /// See the `HDCache` documentation for more information.
+    hd_cache: HDCache,
+}
+
+impl ModuleCache {
+    /// Creates a new `ModuleCache`
+    pub fn new(cfg: ModuleCacheConfig) -> Self {
+        Self {
+            lru_cache: LRUCache::new(cfg.lru_cache_size),
+            hd_cache: HDCache::new(
+                cfg.hd_cache_path.clone(),
+                cfg.hd_cache_size,
+                cfg.snip_amount,
+            ),
+            cfg,
+        }
+    }
+
+    /// Internal function to compile and build `ModuleInfo`
+    fn compile_cached(&mut self, bytecode: &[u8], hash: Hash) -> ModuleInfo {
+        match RuntimeModule::new(
+            bytecode,
+            self.cfg.compilation_gas,
+            self.cfg.gas_costs.clone(),
+            Compiler::CL,
+        ) {
+            Ok(module) => {
+                warn!("compilation of module {} succeeded", hash);
+                ModuleInfo::Module(module)
+            }
+            Err(e) => {
+                warn!("compilation of module {} failed with: {}", hash, e);
+                ModuleInfo::Invalid
+            }
+        }
+    }
+
+    /// Save a new or an already existing module in the cache
+    pub fn save_module(&mut self, bytecode: &[u8]) {
+        let hash = Hash::compute_from(bytecode);
+        if let Some(hd_module_info) =
+            self.hd_cache
```

Comment:
What is this check for ? I assume it's for optimizing and don't try to re-add a module that is already saved in one of the caches but if the module already exist in both then in this case it will fall in this `if` and re-save it in `lru_cache`

Reply:
In a LRU, if you insert an already existing entry it moves it up in the cache